### ISSUE TITLE
A wrapper for optimizers

### ISSUE THREAD

**Issue by jonpsy** [mlpack/mlpack#2969]
### What is the desired addition or change?

Would be nice if we could have wrappers around optimizers. So that we can schedule learning rates and other parameters. For example: Moving Average, Stochastic Moving Average.

### What is the motivation for this feature?

They're pretty cool and allows for a sophisticated deep learning model. Their weight averaging systems acts as an "ensemble" and SWA in particular is specifically known for giving us a "wider" optima range.

### If applicable, describe how this feature would be implemented.

Should be super-easy, barely an inconvenience. Using [tensorflow-addons as inspiration](https://www.tensorflow.org/addons/api_docs/python/tfa/optimizers/MovingAverage), we could accept the internal optimizer as a template argument (with some trait checks)

something like

```c++
template<typename InternalOptimizerType, typename SchecdulePolicy>
typename std::enable_if<traits::IsOptimizerValid<InternalOptimizer>::value, bool>>::value
Optimize( ......)
{
   typename typedef InternalOptimizer::MatType MatType;
   typename typedef InternalOptimizer::GradType GradType;
   
   policy.execute(internalOptimizer);
}
```

#### Shouldn't this be a callback though?
 Both [tensorfow](https://www.tensorflow.org/addons/api_docs/python/tfa/optimizers/MovingAverage)  and [pytorch](https://pytorch.org/blog/stochastic-weight-averaging-in-pytorch/)  has refrained from this approach. Personally, I think this is because callbacks directly interfere with the training process, rather they query useful information from it during the training process. These, however, apply gradient/change weights during the training process and shouldn't be mixed with an array of callbacks.

Let me know what you guys think.

### Additional information?

https://arxiv.org/abs/1803.05407

**Comment by rcurtin** [comment]
> Shouldn't this be a callback though?

I'm pretty sure this is exactly what we designed callbacks for.  If you want to take some action after an epoch or after a step was taken---including modifying the parameters or changing parameters of the optimizer---that's totally reasonable.  Perhaps in TensorFlow or PyTorch, callbacks are meant to be "read-only", but ensmallen isn't PyTorch or TensorFlow and so we don't have the same restrictions.

When I write this response, I'm only thinking of, e.g., modifying the step size / learning rate or changing parameters at the end of an epoch or something.  If you had some other use case in mind, then I've probably overlooked that, but at least for the cases I'm thinking of, callbacks seem completely sufficient.

**Comment by jonpsy** [comment]
What you've mentioned is correct. However, I think we should define the boundaries of what the callbacks can be used for. In Pytorch's opinion _"Callbacks should capture NON-ESSENTIAL logic that is NOT required for your lightning module to run"_ [link](https://pytorch-lightning.readthedocs.io/en/latest/extensions/callbacks.html), and I, for one, think it makes perfect sense.  Logging the progress of training is one thing but it's a whole another thing to allow the Callbacks to apply the gradient and changing weights themselves (as is the case in Stochastic Weight Averaging).

Of course, here, the scheduling of the learning rate is a bit of gray area. If we go by the definition of "Callbacks should only allow changing of **hyper** parameters and should be **non-essential**" then the case of ```learning_rate``` is covered as well. Since, ```learning_rate``` is a hyperparameter we can still keep it as a callback.

Further, there might be cases in which we might want to apply callbacks atop these "special optimizers". This can't be cleanly handled if these optimizers were used as callbacks.

**Comment by rcurtin** [comment]
> However, I think we should define the boundaries of what the callbacks can be used for.

Why restrict functionality?  PyTorch is designed differently, and while in their context perhaps restricting to non-essential logic "makes perfect sense", it does not in the context of ensmallen.

I still don't understand the use case you are getting at here or why our existing callback infrastructure is insufficient.

**Comment by zoq** [comment]
> In Pytorch's opinion _"Callbacks should capture NON-ESSENTIAL logic that is NOT required for your lightning module to run"_ [link](https://pytorch-lightning.readthedocs.io/en/latest/extensions/callbacks.html), and I, for one, think it makes perfect sense.

I'm curious to hear why that makes perfect sense to you? TF and PyTorch implement e.g. SWR as an specific optimizer, correct me if I'm wrong, PyTorch Lightning implements it as a callback https://github.com/PyTorchLightning/pytorch-lightning/blob/ad9118f04afe7e9453418adc8298dee0b5868a14/pytorch_lightning/callbacks/stochastic_weight_avg.py#L36.
